### PR TITLE
Scripts/Isle of Quel'Danas: Thalorien Dawnseeker

### LINKED ISSUE #12444: [Quest] Thalorien Dawnseeker
Core revision: 0f0cea1
Database version: TDB 335.53

http://de.wowhead.com/quest=24563
The whole Questevent is missing:
https://www.youtube.com/watch?feature=player_detailpage&v=UFURxcDcTlQ#t=100

This fix in https://github.com/TrinityCore/TrinityCore/issues/2714 is a hack. Actually there must be a spawnevent.
texts: http://www.wowwiki.com/Quest:Thalorien_Dawnseeker_%28Alliance%29
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/3019671-quest-thalorien-dawnseeker?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Sorikoff**
**Changes proposed:** An attempt to implement quest Thalorien Dawnseeker. All events are linked to Thalorien Dawnseekers Remains which are also correctly positioned right now. 

**Target branch(es):** 3.3.5 (can be ported though)

**Tests performed:** Does work.

**Issues addressed:** Closes #12444

**Known issues and TODO list:** (add/remove lines as needed)

- [x] Sunwell Defenders right now have not exactly correct paths. Looks to me each of them must have own waypoints (although they run away anyway, huh...).

### CODE REVIEW

File: src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+                    case EVENT_OUTRO_4:
+                        if (_thalorien)
+                            _thalorien->AI()->Talk(SAY_THALORIEN_10);
+
+                        _events.ScheduleEvent(EVENT_KNEEL, 6s);
+                        break;
+                    case EVENT_KNEEL:
+                        if (_thalorien)
+                            _thalorien->SetStandState(UNIT_STAND_STATE_KNEEL);
+                            _thalorien->DespawnOrUnsummon(5 * IN_MILLISECONDS);
```
Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp]: I think you are missing some brackets { } here

File: src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+                    case EVENT_OUTRO_2:
+                        if (_thalorien)
+                            _thalorien->AI()->Talk(SAY_THALORIEN_8);
+
+                        _events.ScheduleEvent(EVENT_OUTRO_3, 7s);
+                        break;
+                    case EVENT_OUTRO_3:
+                        if (_thalorien)
+                            _thalorien->AI()->Talk(SAY_THALORIEN_9);
+                            _thalorien->AI()->DoCastSelf(SPELL_POLYMORPH_VISUAL);
```
Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp]: I think you are missing some brackets { } here

File: src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+                    case EVENT_MORLEN_2:
+                        if (_morlen)
+                            _morlen->AI()->Talk(SAY_MORLEN_5);
+
+                        _events.ScheduleEvent(EVENT_MORLEN_ATTACK, 1s);
+                        break;
+                    case EVENT_MORLEN_ATTACK:
+                        if (_morlen)
+                            _morlen->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_PC | UNIT_FLAG_IMMUNE_TO_NPC);
+                            if (_thalorien)
```
Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp]: I think you are missing some brackets { } here

File: src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+                    case EVENT_INTRO_5:
+                        if (_thalorien)
+                            _thalorien->AI()->Talk(SAY_THALORIEN_6);
+
+                        _events.ScheduleEvent(EVENT_MORLEN_1, 6s);
+                        break;
+                    case EVENT_SUMMON_MORLEN:
+                        if (Creature* morlen = me->SummonCreature(NPC_MORLEN_GOLDGRIP, morlenSummon, TEMPSUMMON_CORPSE_DESPAWN, 0))
+                            _morlen = morlen;
+                            _morlen->AI()->DoCastSelf(SPELL_BLOOD_PRESENCE);
```
Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp]: I think you are missing some brackets { } here

**Review by jackpoz — CHANGES_REQUESTED** [review]

File: src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+                if (summon->GetEntry() == NPC_MORLEN_GOLDGRIP)
+                    _events.ScheduleEvent(EVENT_OUTRO_1, 2s);
+            }
+        }
+
+    private:
+        EventMap _events;
+        SummonList _summons;
+        ObjectGuid _playerGUID;
+        Creature* _thalorien;
```
[the remark this answers is not in the document]
Comment (Sorikoff) [reply src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp]: Good point. Can you tell me whether replacing pointers to ObjectAccessor's could be a solution?

File: sql/updates/world/3.3.5/9999_99_99_99_world.sql [hunk 6]
```diff
@@ -0,0 +1,111 @@
+-- Phasing
+DELETE FROM `spell_area` WHERE `spell`=70193 AND `area`=4092;
+INSERT INTO `spell_area` (`spell`,`area`,`quest_start`,`quest_end`,`aura_spell`,`racemask`,`gender`,`autocast`,`quest_start_status`,`quest_end_status`) VALUES
+(70193,4092,24535,0,0,0,2,1,10,0),
+(70193,4092,24563,0,0,0,2,1,10,0);
```
Comment (Killyana) [inline sql/updates/world/3.3.5/9999_99_99_99_world.sql]: Does the phasing ends when the quest is rewarded? no nee to set `quest_end_status`=11?

Comment (Sorikoff) [reply sql/updates/world/3.3.5/9999_99_99_99_world.sql]: This spell is also used in the next quest.

File: src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+public:
+    npc_thalorien_dawnseeker() : CreatureScript("npc_thalorien_dawnseeker") { }
+
+    struct npc_thalorien_dawnseekerAI : public ScriptedAI
+    {
+        npc_thalorien_dawnseekerAI(Creature* creature) : ScriptedAI(creature), _summons(me) { }
+
+        bool GossipSelect(Player* player, uint32 menuId, uint32 /*gossipListId*/) override
+        {
+            if (menuId == MENU_EXAMINE_REMAINS)
```
Comment (Riztazz) [inline src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp]: What if two players have gossip window open? and both click it

File: sql/updates/world/3.3.5/9999_99_99_99_world.sql [hunk 8]
```diff
@@ -20,9 +20,9 @@ INSERT INTO `conditions` (`SourceTypeOrReferenceId`,`SourceGroup`,`SourceEntry`,
 (15,10908,0,0,0,9,0,24535,0,0,0,0,0,'','Show gossip option if 24535 is incomplete'),
 (15,10908,0,0,1,9,0,24563,0,0,0,0,0,'','Show gossip option if 24563 is incomplete');
 
-DELETE FROM `gameobject` WHERE `id`=201718;
+DELETE FROM `gameobject` WHERE `guid`=9004;
```
Comment (jackpoz) [inline sql/updates/world/3.3.5/9999_99_99_99_world.sql]: please delete by guid AND id, just to make sure we are deleting exactly what we want and not some other gameobject

File: src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp (hunk 7 above)
Comment (Sorikoff) [reply src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp]: After first one clicks, menu is closed and event begins.

Comment (Riztazz) [reply src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp]: Is it closed for the other player tho?

Comment (Sorikoff) [reply src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp]: Closed for both meaning you won't be able to start it again shortly.

File: src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+                    _thalorienGUID = thalorien->GetGUID();
+                me->SummonCreatureGroup(SUMMON_SUNWELL_DEFENDER);
+
+                _events.ScheduleEvent(EVENT_INTRO_0, 3s);
+            }
+        }
+
+        void Reset() override
+        {
+            me->SetFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_GOSSIP);
```
Comment (Killyana) [inline src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp]: Add the npc flag on the DB

Comment (Sorikoff) [reply src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp]: It's already set, isn't it?
```sql
mysql> SELECT `npcflag` FROM `creature_template` WHERE `entry`=37552;
+---------+
| npcflag |
+---------+
|       1 |
+---------+
1 row in set (0.00 sec)
```

File: sql/updates/world/3.3.5/9999_99_99_99_world.sql [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+(70193,4092,24535,0,0,0,2,1,10,0),
+(70193,4092,24563,0,0,0,2,1,10,0);
+
+-- Thalorien Dawnseekers Remains
+UPDATE `creature` SET `phaseMask`=2049,`position_x`=11792.01,`position_y`=-7065.793,`position_z`=25.94043,`orientation`=5.009095 WHERE `guid`=950;
+UPDATE `creature_template` SET `gossip_menu_id`=10908,`ScriptName`='npc_thalorien_dawnseeker' WHERE `entry`=37552;
+DELETE FROM `creature_template_addon` WHERE `entry`=37552;
+INSERT INTO `creature_template_addon` (`entry`,`bytes2`,`auras`) VALUES (37552,1,'29266');
+
+DELETE FROM `gossip_menu` WHERE `MenuID`=10908;
```
Comment (Killyana) [inline sql/updates/world/3.3.5/9999_99_99_99_world.sql]: Remove the gossip 37552 as it's wrong
``` sql
DELETE FROM `gossip_menu` WHERE `MenuID` IN (10908,37552 );

File: src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sorikoff) [reply src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp]: Yeah it's fine. I was afraid of performance, but profiler didn't confirm that.

**Comment by Sorikoff** [comment]
_An accident happened._ Sorry.

Superseded by #23382.

**Pull request closed without merging**